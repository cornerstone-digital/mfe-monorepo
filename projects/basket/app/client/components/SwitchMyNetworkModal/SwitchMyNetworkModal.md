The `SwitchMyNetworkModal` organism houses a form used for customers to enter their PAC or STAC code and mobile number, validate it, select a date to port their number over to Vodafone and save it to the basket. It's used in the basket to allow customers to enter their PAC or STAC at the point of sale.

```
initialState = {open: false}
;<div>
  <a style={{textDecoration: 'underline', cursor: 'pointer'}} onClick={() => { setState({open: true})}}>
    Open SwitchMyNetworkModal in its default state
  </a>
  <SwitchMyNetworkModal isOpen={state.open} onExit={() => { setState({open: false})}} />
</div>
```

If the delivery date the customer has in the checkout is later than the porting or termination date selected by the customer in the basket, then they can show a different state of this organism to inform them of this, using the `hasLateDelivery` boolean prop. A `phoneNumber` and `date` is expected from the `portability` prop, which will render in the copy and is the details relevant to the customer. Additionally, a `codeType` should also be present in `portabiliy` — PAC or STAC — that tweaks the copy to be relevant to the type of switch they are doing. You can provide an `onContinue` function, which will be called after they've successfully clicked 'Continue'. The 'onContinue' function should ideally set the `isOpen` prop to false in order to exit the component.

```
initialState = {open: false}
;<div>
  <a style={{textDecoration: 'underline', cursor: 'pointer'}} onClick={() => { setState({open: true})}}>
    Open SwitchMyNetworkModal in its late delivery state
  </a>
  <SwitchMyNetworkModal isOpen={state.open} onExit={() => { setState({open: false})}} portability={{codeType: 'PAC', msisdn: '07123456789', validPortDate: '01/02/2020' }} hasLateDelivery={true} />
</div>
```

Similar to the above, if the customer has reached the checkout, but the package for which they have filled out a PAC or STAC code in the basket cannot be delivered before the date they selected for switching due to a backorder/preorder, the they will see the same scenario (albeit slightly different copy). This should be provided to SwitchMyNetworkModal as `hasUnknownDelivery`, as the delivery date is not known at this stage and cannot be guaranteed. This just alters copy shown to the customer, but is functionally the same as above. You can provide an `onContinue` function again.

```
initialState = {open: false}
;<div>
  <a style={{textDecoration: 'underline', cursor: 'pointer'}} onClick={() => { setState({open: true})}}>
    Open SwitchMyNetworkModal in its unknown preorder/backorder delivery state
  </a>
  <SwitchMyNetworkModal isOpen={state.open} onExit={() => { setState({open: false})}} portability={{codeType: 'PAC', msisdn: '07123456789', validPortDate: '01/02/2020' }} hasUnknownDelivery={true} />
</div>
```
