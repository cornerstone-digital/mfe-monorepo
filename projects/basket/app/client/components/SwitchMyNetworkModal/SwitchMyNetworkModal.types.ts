export interface SwitchMyNetworkModalProps {
  /** basket for analytics */
  basket: BasketV2.Basket
  /** the ID of the package the switch my network applies to */
  packageId?: string
  /** the date of delivery, which can be used in the copy if needed */
  deliveryDate?: string
  /** alters content to show scenario when delivery is after port-in/termination date */
  hasLateDelivery?: boolean
  /** alters content to show scenario when delivery is unknown due to backorder or other */
  hasUnknownDelivery?: boolean
  /** function to call when organism is finished with (e.g. to close modal) */
  onExit: () => void

  /** the portability info filled out by the user from the basket */
  portability: BasketV2.Portability
  /** whether the SwitchMyNetwork modal is open */
  isOpen?: boolean
  onUpdateBasket: () => Promise<void>
  pageError?: string
  reviewMode?: boolean
}
