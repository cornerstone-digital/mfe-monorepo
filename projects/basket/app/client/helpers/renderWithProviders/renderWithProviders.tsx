import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import theme from '@vfuk/core-theme-ws2'
import { OverlayProvider } from '@vfuk/core-overlay-controller'

const renderWithProviders = (Component: React.ReactElement) => {
  const wrapper = document.createElement('div')
  wrapper.id = 'GlobalStylesProvider'
  return render(
    <ThemeProvider theme={theme}>
      <OverlayProvider>{Component}</OverlayProvider>
    </ThemeProvider>,
    { container: document.body.appendChild(wrapper) },
  )
}

export default renderWithProviders
