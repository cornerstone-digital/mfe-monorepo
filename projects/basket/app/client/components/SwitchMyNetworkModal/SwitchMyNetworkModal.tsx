import Paragraph from '@web-core/components/atoms/Paragraph'

import Modal from '@vfuk/core-modal'
import Heading from '@web-core/components/atoms/Heading'
import basketService from '@web-shop-core/services/basketService'

import AnalyticsUtil from '@utilities/Analytics'

import FindOutMore from './components/FindOutMore'
import PacStacExplained from './components/PacStacExplained'
import SwitchMyNetworkForm from './components/SwitchMyNetworkForm'
import ContinueWithoutSwitching from './components/ContinueWithoutSwitching'

import { SwitchMyNetworkModalProps } from './SwitchMyNetworkModal.types'

import styles from './SwitchMyNetworkModal.scss'

const SwitchMyNetworkModal = (props: SwitchMyNetworkModalProps) => {
  const {
    hasLateDelivery,
    hasUnknownDelivery,
    packageId,
    onUpdateBasket,
    reviewMode,
    basket,
    pageError,
    isOpen,
    onExit,
    portability,
    deliveryDate,
  } = props
  const { basketId = '' } = basket

  const title = hasLateDelivery
    ? 'Choose a different transfer date'
    : hasUnknownDelivery
    ? `We can't guarantee that transfer date`
    : 'Switch My Network'
  const showContinueWithoutSwitching = hasLateDelivery || hasUnknownDelivery

  const handleSave = async (portabilityInfo: BasketV2.Portability = {}) => {
    await basketService.submitPortabilityInfo(basketId, packageId, portabilityInfo)

    if (onUpdateBasket) {
      await onUpdateBasket()

      if (!reviewMode) {
        const codeType = portabilityInfo.codeType && portabilityInfo.codeType.toLowerCase()
        AnalyticsUtil.trackLink('basketPage.switchNetworkSuccess', {
          newBasket: basket,
          packageId,
          codeType,
          pageError,
        })
      }
    }
  }

  const handleRemove = async () => {
    await basketService.removePortabilityInfo(basketId, packageId)

    if (onUpdateBasket) {
      await onUpdateBasket()
    }
  }

  const handleContinue = async (attemptRemoval?: boolean) => {
    if (attemptRemoval) {
      await handleRemove()
    }
  }

  return (
    <Modal isOpen={!!isOpen} srName={title} onCloseCb={onExit} size={2}>
      <>
        <Heading level={3} size={3}>
          {title}
        </Heading>
        <div className={styles['wrapper']}>
          {showContinueWithoutSwitching ? (
            <ContinueWithoutSwitching
              onContinue={handleContinue}
              phoneNumber={portability?.msisdn}
              date={deliveryDate}
              hasLateDelivery={hasLateDelivery}
              hasUnknownDelivery={hasUnknownDelivery}
            />
          ) : (
            <>
              <Paragraph>
                You can bring your current mobile number with you when you switch to Vodafone, or get a new number. This can be done now, or
                after you&apos;ve finished your purchase.
              </Paragraph>
              <Paragraph>Enter a PAC or STAC code and choose a date to get started.</Paragraph>
              <PacStacExplained />
              <SwitchMyNetworkForm
                basket={basket}
                onExit={onExit}
                onSave={handleSave}
                onRemove={handleRemove}
                portability={portability}
                packageId={packageId}
                reviewMode={reviewMode}
              />
              <FindOutMore />
            </>
          )}
        </div>
      </>
    </Modal>
  )
}

export default SwitchMyNetworkModal
