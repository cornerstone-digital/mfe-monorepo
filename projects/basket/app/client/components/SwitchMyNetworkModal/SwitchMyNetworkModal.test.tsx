import { fireEvent, waitFor } from '@testing-library/react'

import basketService from '@web-shop-core/services/basketService'
import AnalyticsUtil from '@utilities/Analytics'

import SwitchMyNetworkModal from './SwitchMyNetworkModal'
import renderWithProviders from '@helpers/renderWithProviders'

jest.mock('../../utilities/Analytics')
jest.mock('@web-shop-core/services/basketService')
jest.mock('@vfuk/core-modal')

AnalyticsUtil.trackLink = jest.fn()
basketService.submitPortabilityInfo = jest.fn()

const mockBasketId = 'mock-basket-id'
const mockPackageId = 'mock-package-id'

const props = {
  basket: { basketId: mockBasketId },
  packageId: 'package-id',
  portability: {},
  pageError: '',
  reviewMode: false,
  isOpen: true,
  onExit: jest.fn(),
  onContinue: jest.fn(),
  onUpdateBasket: jest.fn(),
}

describe('<SwitchMyNetworkModal />', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  afterEach(() => {
    jest.clearAllMocks()
    jest.restoreAllMocks()
  })

  it('should render without exploding', () => {
    const wrapper = shallow(<SwitchMyNetworkModal {...props} />)
    expect(snapshot(wrapper)).toMatchSnapshot()
  })

  describe('Modal', () => {
    it('should be visible if isOpen is true', async () => {
      const { getByText } = renderWithProviders(<SwitchMyNetworkModal {...props} />)
      await waitFor(() => {
        expect(getByText('Switch My Network')).toBeInTheDocument()
      })
    })
    it('should not be visible if isOpen is false', async () => {
      const { queryByText, container } = renderWithProviders(<SwitchMyNetworkModal {...props} isOpen={false} />)
      await waitFor(() => {
        expect(container.childNodes.length).toBe(0)
        expect(queryByText('Switch My Network')).not.toBeInTheDocument()
      })
    })
  })

  it('should display default modal title', () => {
    const { getByText } = renderWithProviders(<SwitchMyNetworkModal {...props} />)
    expect(getByText('Switch My Network')).toBeInTheDocument()
  })

  it('should display modal title regarding late delivery', () => {
    const { getByText } = renderWithProviders(<SwitchMyNetworkModal {...props} hasLateDelivery />)
    expect(getByText('Choose a different transfer date')).toBeInTheDocument()
  })

  it('should display modal title regarding unknown delivery', () => {
    const { getByText } = renderWithProviders(<SwitchMyNetworkModal {...props} hasUnknownDelivery />)
    expect(getByText("We can't guarantee that transfer date")).toBeInTheDocument()
  })

  it('should display <ContinueWithoutSwitching /> if it has late delivery', () => {
    const wrapper = shallow(<SwitchMyNetworkModal {...props} hasLateDelivery />)
    expect(wrapper.find('ContinueWithoutSwitching')).toHaveLength(1)
  })

  it('should display <ContinueWithoutSwitching /> if it has unknown delivery', () => {
    const wrapper = shallow(<SwitchMyNetworkModal {...props} hasUnknownDelivery />)
    expect(wrapper.find('ContinueWithoutSwitching')).toHaveLength(1)
  })

  describe('handleExit', () => {
    it('should call onExit when clicking on close button', async () => {
      const onExitSpy = jest.fn()
      const { container } = renderWithProviders(<SwitchMyNetworkModal {...props} isOpen onExit={onExitSpy} />)
      const button = container.querySelector("button[aria-label='close']") as HTMLButtonElement
      await waitFor(() => {
        expect(button).toBeInTheDocument()
      })
      fireEvent.click(button)
      expect(onExitSpy).toBeCalled()
    })
  })

  describe('handleSave', () => {
    it('should call corresponding service when form is saved', () => {
      const wrapper = shallow(<SwitchMyNetworkModal {...props} packageId={mockPackageId} />)
      wrapper.find('SwitchMyNetworkForm').prop('onSave')()
      expect(basketService.submitPortabilityInfo).toBeCalledWith(mockBasketId, mockPackageId, {})
    })

    it('should call onUpdateBasket if provided', async () => {
      const mockOnUpdateBasket = jest.fn(() => Promise.resolve())
      const wrapper = shallow(<SwitchMyNetworkModal {...props} onUpdateBasket={mockOnUpdateBasket} reviewMode />)
      await wrapper.find('SwitchMyNetworkForm').prop('onSave')()
      expect(mockOnUpdateBasket).toBeCalled()
    })

    it('should call analytics if onUpdateBasket is provided and reviewMode is false', async () => {
      const mockOnUpdateBasket = jest.fn(() => Promise.resolve())
      const wrapper = shallow(<SwitchMyNetworkModal {...props} onUpdateBasket={mockOnUpdateBasket} />)
      await wrapper.find('SwitchMyNetworkForm').prop('onSave')()
      expect(basketService.submitPortabilityInfo).toBeCalled()
    })
  })

  describe('handleRemove', () => {
    it('should call corresponding service when removing form', async () => {
      const wrapper = shallow(<SwitchMyNetworkModal {...props} packageId={mockPackageId} />)
      await wrapper.find('SwitchMyNetworkForm').prop('onRemove')()
      expect(basketService.removePortabilityInfo).toBeCalledWith(mockBasketId, mockPackageId)
    })

    it('should call onUpdateBasket if provided', async () => {
      const mockOnUpdateBasket = jest.fn(() => Promise.resolve())
      const wrapper = shallow(<SwitchMyNetworkModal {...props} onUpdateBasket={mockOnUpdateBasket} />)
      await wrapper.find('SwitchMyNetworkForm').prop('onRemove')()
      expect(mockOnUpdateBasket).toBeCalled()
    })
  })

  describe('handleContinue', () => {
    it('should call handleRemove if attemptRemoval is true', async () => {
      const wrapper = shallow(<SwitchMyNetworkModal {...props} hasLateDelivery packageId={mockPackageId} />)
      await wrapper.find('ContinueWithoutSwitching').prop('onContinue')(true)
      expect(basketService.removePortabilityInfo).toBeCalledWith(mockBasketId, mockPackageId)
    })
  })
})
