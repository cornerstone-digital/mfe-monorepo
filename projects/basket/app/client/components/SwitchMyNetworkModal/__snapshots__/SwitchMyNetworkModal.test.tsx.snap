// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SwitchMyNetworkModal /> should render without exploding 1`] = `
<Modal
  isOpen={true}
  onCloseCb={[MockFunction]}
  size={2}
  srName="Switch My Network"
>
  <Heading
    appearance="inherit"
    focusable={false}
    level={3}
    size={3}
    theme={null}
  >
    Switch My Network
  </Heading>
  <div
    className="wrapper"
  >
    <Paragraph
      theme={null}
    >
      You can bring your current mobile number with you when you switch to Vodafone, or get a new number. This can be done now, or after you've finished your purchase.
    </Paragraph>
    <Paragraph
      theme={null}
    >
      Enter a PAC or STAC code and choose a date to get started.
    </Paragraph>
    <PacStacExplained />
    <SwitchMyNetworkForm
      basket={
        Object {
          "basketId": "mock-basket-id",
        }
      }
      onExit={[MockFunction]}
      onRemove={[Function]}
      onSave={[Function]}
      packageId="package-id"
      portability={Object {}}
      reviewMode={false}
    />
    <FindOutMore />
  </div>
</Modal>
`;
